src/Avalonia.Layout/LayoutManager.cs

```diff
@@ -202,6 +196,49 @@ public void Dispose()
             _toArrange.Dispose();
         }
 
+        void ILayoutManager.RegisterEffectiveViewportListener(ILayoutable control)
+        {
+            _effectiveViewportChangedListeners ??= new List<EffectiveViewportChangedListener>();
+            _effectiveViewportChangedListeners.Add(new EffectiveViewportChangedListener(
+                control,
+                CalculateEffectiveViewport(control)));
+        }
+
+        void ILayoutManager.UnregisterEffectiveViewportListener(ILayoutable control)
+        {
+            if (_effectiveViewportChangedListeners is object)
+            {
+                for (var i = 0; i < _effectiveViewportChangedListeners.Count; ++i)
```

Comment:
You should iterate backwards to prevent issues when more than one listener is present.

Reply:
Hmm, I did at one point add a TODO comment there to duplicate the listeners before iterating. Not sure iterating backwards would help much because a notification can result in arbitrary listeners being removed.